File: apps/cargo-tracker/src/contexts/billing/domain/model/payment-status.ts
/** 支払い状態（PENDING / CONFIRMED / OVERDUE / REFUNDED） */
export const PaymentStatus = {
  PENDING: 'PENDING',
  CONFIRMED: 'CONFIRMED',
  OVERDUE: 'OVERDUE',
  REFUNDED: 'REFUNDED',
} as const;
export type PaymentStatus = (typeof PaymentStatus)[keyof typeof PaymentStatus];

export const PAYMENT_STATUS_LABELS: Record<PaymentStatus, string> = {
  PENDING: '未払い',
  CONFIRMED: '入金確認済',
  OVERDUE: '支払期限超過',
  REFUNDED: '返金済',
};

export function isPaymentStatus(value: string): value is PaymentStatus {
  return Object.values(PaymentStatus).includes(value as PaymentStatus);
}

File: docs/article/source/ruby/apps/cargo-tracker/packs/billing/app/domain/billing/domain/payment_status.rb
# frozen_string_literal: true

module Billing
  module Domain
    # 支払状態（PaymentStatus）。請求書の精算ライフサイクルを表す不変値オブジェクト（US23）。
    class PaymentStatus
      PENDING = "PENDING"
      CONFIRMED = "CONFIRMED"
      OVERDUE = "OVERDUE"
      REFUNDED = "REFUNDED"

      VALUES = [ PENDING, CONFIRMED, OVERDUE, REFUNDED ].freeze

      attr_reader :value

      def self.initial = new(value: PENDING)

      def initialize(value:)
        raise ArgumentError, "支払状態が不正です: #{value}" unless VALUES.include?(value)

        @value = value
        freeze
      end

      def pending? = value == PENDING
      def confirmed? = value == CONFIRMED
      def overdue? = value == OVERDUE
      # 未精算（入金確認・料金調整の対象・PENDING or OVERDUE）。
      def unsettled? = pending? || overdue?

      def ==(other) = other.is_a?(PaymentStatus) && other.value == value
      alias eql? ==
      def hash = value.hash
      def to_s = value
    end
  end
end
